R/oligoNucleotideEnrichment.R

```diff
@@ -0,0 +1,188 @@
+#'
+#' @details
+#'
+#' For the paramter k, method, nthreads, window, window.size and window.overlap, please see \link[universalmotif]{shuffle_sequences} for more information.
+#'
+#' @author Junhui Li
+#'
+#' @import  Biostrings
+#' 
+#' @importFrom stats binom.test
+#' 
+#' @importFrom universalmotif shuffle_sequences
+#'
+#' @examples
+#' library(BSgenome.Hsapiens.UCSC.hg19)
+#' filepath =system.file("extdata", "examplePattern.fa", package="ChIPpeakAnno")
+#' peaks = GRanges(seqnames=c("chr17","chr3","chr12","chr8"),
+#'                  IRanges(start=c(41275784,10076141,4654135,31024288),
+#'                          end=c(41276382,10076732,4654728,31024996),
+#'                          names=paste0("peak",1:4)))
+#' \dontrun{
+#' result <- oligoNucleotideEnrichment(filepath=filepath,
+#' peaks=peaks,
+#' genome=Hsapiens,
+#' background="randomPick")
+#' }
+#' @export
+
+oligoNucleotideEnrichment <- function(filepath,
+                                      format="fasta",
+                                      peaks,
+                                      genome,
+                                      background=c("randomPick","shuffle"),
+                                      chromosome=NULL,
+                                      ...,
+                                      times=1000,
+                                      alpha=0.05){
+  if(!file.exists(filepath)){
+    stop("file doesn't exist!")
+  }
+  background <- match.arg(background)
+  method <- match.arg(method)
+  if(!is(genome, "BSgenome") & !is(genome, "Mart") | missing(genome)){
+    stop("genome is required parameter, 
+             please pass in either a BSgenome object or a Mart object!")
+  }
+  if(round(times*alpha,0)==0){
+    stop("The paramter 'times' should be increased to large enough!")
+  }
+  nPeak <- length(peaks)
+  allseqLen <- seqlengths(genome)
+  seqWidth <- width(peaks)
+  candiChrom <- names(allseqLen >= max(seqWidth))
+  if(any(seqWidth > allseqLen[as.character(seqnames(peaks)@values)])){
+    stop("The length of sequence should be less than the length of chromosomes")
+  }
+  actualSeqs <- getAllPeakSequence(peaks, upstream = 0, downstream = 0, genome = genome)
+  actualP <- oligoNucleotideSummary(filepath,format=format,seqs=actualSeqs)
+  
+  statDist <- matrix(NA,0,4)
+  colnames(statDist) <- c("x","n","prop.background","binom.pvalue")
+  if(background=="randomPick"){
+    statDist <- do.call(rbind,lapply(seq.int(times),function(n) {
+      if(is.null(chromosome)==TRUE){
+        chrs <- as.character(seqnames(peaks))
+        givenSeqLen <- allseqLen[as.character(seqnames(peaks))]
+      }else{
+        chrs <- sample(candiChrom,nPeak)
+        givenSeqLen <- allseqLen[chrs]
+      }
+      startPos <- sapply(givenSeqLen-seqWidth,function(x){sample(seq.int(x),1)})
+      endPos <- startPos + seqWidth - 1
```

Comment:
bug here. If the length(nPeak)==0, it will raise a error.

Reply:
added `stopifnot` in previous part.